File: cpp/code_forces/edu/itmo/seg_tree1/part4/2.py
class SegmentTree:
    def __init__(self, arr, mod):
        self.n = len(arr)
        self.mod = mod
        self.tree = [0] * (4 * self.n)
        self.lazy = [None] * (4 * self.n)
        self.build(arr, 1, 0, self.n - 1)

    def build(self, arr, v, tl, tr):
        if tl == tr:
            self.tree[v] = arr[tl]
        else:
            tm = (tl + tr) // 2
            self.build(arr, 2 * v, tl, tm)
            self.build(arr, 2 * v + 1, tm + 1, tr)
            self.tree[v] = self.multiply(self.tree[2 * v], self.tree[2 * v + 1])

    def multiply(self, a, b):
        c = [[0, 0], [0, 0]]
        for i in range(2):
            for j in range(2):
                for k in range(2):
                    c[i][j] += a[i][k] * b[k][j]
                    c[i][j] %= self.mod
        return c

    def update(self, v, tl, tr, pos, new_val):
        if tl == tr:
            self.tree[v] = new_val
        else:
            tm = (tl + tr) // 2
            if pos <= tm:
                self.update(2 * v, tl, tm, pos, new_val)
            else:
                self.update(2 * v + 1, tm + 1, tr, pos, new_val)
            self.tree[v] = self.multiply(self.tree[2 * v], self.tree[2 * v + 1])

    def query(self, v, tl, tr, l, r):
        if l > r:
            return [[1, 0], [0, 1]]
        if l == tl and r == tr:
            return self.tree[v]
        tm = (tl + tr) // 2
        left_query = self.query(2 * v, tl, tm, l, min(r, tm))
        right_query = self.query(2 * v + 1, tm + 1, tr, max(l, tm + 1), r)
        return self.multiply(left_query, right_query)

# Чтение входных данных
r, n, m = map(int, input().split())
matrices = []
for _ in range(n):
    matrix = [list(map(int, input().split())) for _ in range(2)]
    matrices.append(matrix)

# Создание и инициализация дерева отрезков
seg_tree = SegmentTree(matrices, r)

# Обработка запросов
for _ in range(m):
    l, r = map(int, input().split())
    result = seg_tree.query(1, 0, n - 1, l - 1, r - 1)
    for row in result:
        print(*row)
    print()

File: cpp/code_forces/edu/itmo/seg_tree1/part4/2.cpp
#include <bits/stdc++.h>
using namespace std;

long long MOD = 0;

struct segtree
{
    struct mat
    {
        long long a, b, c, d;
    };

    vector<mat> tree;
    int size;

    const mat ZERO{1, 0, 0, 1};

    segtree(int n)
    {
        size = n;
        tree.assign(4 * size, ZERO);
        build(1, 0, size - 1);
    }

    void build(int x, int tl, int tr)
    {
        if (tl == tr)
        {
            mat m = ZERO;
            cin >> m.a >> m.b >> m.c >> m.d;
            tree[x] = m;
            return;
        }
        int tm = (tl + tr) / 2;
        build(x * 2, tl, tm);
        build(x * 2 + 1, tm + 1, tr);
        tree[x] = combine(tree[x * 2], tree[x * 2 + 1]);
    }

    mat combine(mat &x, mat &y)
    {
        mat m{
            ((x.a % MOD * y.a % MOD) % MOD + (x.b % MOD * y.c % MOD) % MOD) % MOD,
            ((x.a % MOD * y.b % MOD) % MOD + (x.b % MOD * y.d % MOD) % MOD) % MOD,
            ((x.c % MOD * y.a % MOD) % MOD + (x.d % MOD * y.c % MOD) % MOD) % MOD,
            ((x.c % MOD * y.b % MOD) % MOD + (x.d % MOD * y.d % MOD) % MOD) % MOD};
        return m;
    }

    mat query(int l, int r)
    {
        return query(l, r, 1, 0, size - 1);
    }

    mat query(int l, int r, int x, int tl, int tr)
    {
        if (l > r)
        {
            return ZERO;
        }
        if (l == tl && r == tr)
        {
            return tree[x];
        }
        int tm = (tl + tr) / 2;
        auto lm = query(l, min(r, tm), x * 2, tl, tm);
        auto rm = query(max(tm + 1, l), r, x * 2 + 1, tm + 1, tr);
        return combine(lm, rm);
    }
};

int main()
{
    ios::sync_with_stdio(false);
    int n, m;
    cin >> MOD >> n >> m;
    segtree t(n);

    while (m-- > 0)
    {
        int l, r;
        cin >> l >> r;
        l--, r--;
        segtree::mat res = t.query(l, r);
        cout << res.a << " " << res.b << "\n";
        cout << res.c << " " << res.d << "\n";
    }
    cout << endl;
    return 0;
}